src/NumberTheory/vinberg.jl

```diff
@@ -0,0 +1,291 @@
+  S = snf(Q)
+  return (-1) * divisors(2 * S[l, l])
+end
+
+@doc raw"""
+    _check_root_lengths(Q::ZZMatrix, root_lengths::Vector{ZZRingElem}) -> Vector{ZZRingElem}
+
+Check whether the given lengths are possible root lengths.
+Unnecessary roots are sorted out and reported if this is wanted.
+Return only the possible root lengths.
+"""
+function _check_root_lengths(Q::ZZMatrix, root_lengths::Vector{ZZRingElem})
+  l = nrows(Q)
+  S = snf(Q)
+  possible_root_lengths = divisors(2 * S[l, l])
+  if isempty(root_lengths)
+    return possible_root_lengths
+  end
+  result = ZZRingElem[]
+  for r in root_lengths
+    if abs(r) in possible_root_lengths
+      push!(result, r)
+    else
+      @vprintln :Vinberg 1 "$r is not a possible root length"
+    end
+  end
+  @req !isempty(result) "No possible root lengths found"
+  return result
+end
+
+@doc raw"""
+    _distance_indices(upper_bound::ZZRingElem, root_lengths::Vector{ZZRingElem}) -> Vector{Tuple{Int, ZZRingElem}}
+
+Returns all possible tuples $(n, l)$ with `n` a natural number and `l` contained in `root_lengths`,
+sorted by increase of the value $\frac{n^2}{l}$, stopping by `upper_bound`.
+"""
+function _distance_indices(upper_bound::Rational, root_lengths::Vector{ZZRingElem})
+
+  result = Tuple{Int,ZZRingElem}[]
+
+  for l in root_lengths
+    x = isqrt(x0*upper_bound) # consider n^2//l < upper_bound => n < sqrt(l*upper_bound)
+    for n in 1:x
+      push!(result, (n, l))
+    end
+  end
+
+  sort!(result; by=x -> (x[1]^2) // abs(x[2]))
+  return result
+end
+
+@doc raw"""
+    _is_primitive(v)
+
+Check if `v` is primitive, which is equivalent to checking whether the greatest common divisor of the entries of `v`
+is equal to 1. If `v` is not primitive, we divide `v` by its gcd.
+"""
+function _is_primitive(v)
+  gcd = reduce(gcd, v)
+  if isone(gcd)
+    return v
+  else 
+    return (1//gcd * v)
+  end
+end
+
+@doc raw"""
+    _crystallographic_condition(Q::ZZMatrix, v::ZZMatrix)
+
+Check if reflection by `v` preserves the lattice, i.e. check if for `v` a row vector
+$\frac{2.v.Q}{v^2}$ is an integer matrix. 
+"""
+function _crystallographic_condition(Q::ZZMatrix, v::ZZMatrix)
+  A = Q * transpose(v)
+  b = (v*A)[1, 1]
+  A = 2 * A
+
+  return all(iszero(mod(x, b)) for x in A)
+end
+
+@doc raw"""
+    _check_coorientation(Q::ZZMatrix, roots::Vector{ZZMatrix}, v::ZZMatrix, v0::ZZMatrix)
+
+First check whether `v` has non-obtuse angles with all roots `r` already found by checking that $v.r \geq 0$.
+Then also check the coorientation of `v` by checking that $v.v_0 \geq 0$.
+"""
+function _check_coorientation(Q::ZZMatrix, roots::Vector{ZZMatrix}, v::ZZMatrix, v0::ZZMatrix)
+  Qv = Q * transpose(v)
+  for r in roots
+    if (r*Qv)[1, 1] < 0
+      return false
+    end
+  end
+
+  return (v*Q*transpose(v0))[1, 1] > 0
+end
+
+@doc raw"""
+    _distance_0(Q::ZZMatrix, v0::ZZMatrix, root_lengths::Vector{ZZRingElem}) -> Vector{ZZMatrix}
+
+Return the roots which are orthogonal to `v0`.
```

Comment:
```suggestion
Return the fundamental roots which are orthogonal to `v0`.
```
A vector `r` in `L` is called a root if $(r.r) \neq 0$ and the reflection in `r` preserves `L`. 
It seems that you are looking for fundamental roots instead.

Reply:
Please state what the direction vector is and that it equals `v1` in your notation.